NPE in SortingCodecReader.getVectorReader()...checkIntegrity()

### Description

I'm confused about this one -- it seems straightforward; the test creates an index with a segment that has no vectors, so its vectorsReader is null, but we assume it is not null when wrapping it in `SortingCodecReader`. So we can easily enough add protection ` (if (null) return null)` in `SortingCodecReader.getVectorReader()` and it is clear that we already expect to handle a null return in the merge call. Indeed I tried this and it fixes the test.

What I'm struggling with is (1) why wasn't this a problem in the past? and (2) Why do we need this protection for KnnVectorsReader, but not for all the other reader types. For example, looking at `SortingCodecReader.getPointsReader()` it just assumes that its delegate reader is non-null.

Re: (1)

```
java.lang.NullPointerException: Cannot invoke "org.apache.lucene.codecs.KnnVectorsReader.checkIntegrity()" because "this.val$delegate" is null

Stacktrace

java.lang.NullPointerException: Cannot invoke "org.apache.lucene.codecs.KnnVectorsReader.checkIntegrity()" because "this.val$delegate" is null
	at org.apache.lucene.index.SortingCodecReader$5.checkIntegrity(SortingCodecReader.java:557)
	at org.apache.lucene.codecs.KnnVectorsWriter.merge(KnnVectorsWriter.java:95)
	at org.apache.lucene.index.SegmentMerger.mergeVectorValues(SegmentMerger.java:271)
	at org.apache.lucene.index.SegmentMerger.mergeWithLogging(SegmentMerger.java:314)
	at org.apache.lucene.index.SegmentMerger.merge(SegmentMerger.java:158)
	at org.apache.lucene.index.IndexWriter.addIndexesReaderMerge(IndexWriter.java:3468)
	at org.apache.lucene.index.IndexWriter$AddIndexesMergeSource.merge(IndexWriter.java:3345)
	at org.apache.lucene.index.ConcurrentMergeScheduler.doMerge(ConcurrentMergeScheduler.java:668)
	at org.apache.lucene.index.ConcurrentMergeScheduler$MergeThread.run(ConcurrentMergeScheduler.java:729)

Standard Error

Jan 09, 2025 8:54:46 PM com.carrotsearch.randomizedtesting.RandomizedRunner$QueueUncaughtExceptionsHandler uncaughtException
WARNING: Uncaught exception in thread: Thread[#695,Lucene Merge Thread #0,5,TGRP-TestSortingCodecReader]
org.apache.lucene.index.MergePolicy$MergeException: java.lang.NullPointerException: Cannot invoke "org.apache.lucene.codecs.KnnVectorsReader.checkIntegrity()" because "this.val$delegate" is null
	at __randomizedtesting.SeedInfo.seed([C61D1CC67658F8FD]:0)
	at org.apache.lucene.index.ConcurrentMergeScheduler.handleMergeException(ConcurrentMergeScheduler.java:764)
	at org.apache.lucene.index.ConcurrentMergeScheduler$MergeThread.run(ConcurrentMergeScheduler.java:756)
Caused by: java.lang.NullPointerException: Cannot invoke "org.apache.lucene.codecs.KnnVectorsReader.checkIntegrity()" because "this.val$delegate" is null
	at org.apache.lucene.index.SortingCodecReader$5.checkIntegrity(SortingCodecReader.java:557)
	at org.apache.lucene.codecs.KnnVectorsWriter.merge(KnnVectorsWriter.java:95)
	at org.apache.lucene.index.SegmentMerger.mergeVectorValues(SegmentMerger.java:271)
	at org.apache.lucene.index.SegmentMerger.mergeWithLogging(SegmentMerger.java:314)
	at org.apache.lucene.index.SegmentMerger.merge(SegmentMerger.java:158)
	at org.apache.lucene.index.IndexWriter.addIndexesReaderMerge(IndexWriter.java:3468)
	at org.apache.lucene.index.IndexWriter$AddIndexesMergeSource.merge(IndexWriter.java:3345)
	at org.apache.lucene.index.ConcurrentMergeScheduler.doMerge(ConcurrentMergeScheduler.java:668)
	at org.apache.lucene.index.ConcurrentMergeScheduler$MergeThread.run(ConcurrentMergeScheduler.java:729)
```


### Gradle command to reproduce

    ./gradlew :lucene:core:test --tests "org.apache.lucene.index.TestSortingCodecReader.testSortOnAddIndicesRandom" -Dtests.seed=C61D1CC67658F8FD -Dtests.multiplier=3

--- a/lucene/CHANGES.txt
+++ b/lucene/CHANGES.txt
@@ -66,6 +66,8 @@ Bug Fixes
 * GITHUB#14109: prefetch may select the wrong memory segment for
   multi-segment slices. (Chris Hegarty)
 
+* GITHUB#14123: SortingCodecReader NPE when segment has no (points, vectors, etc...) (Mike Sokolov)
+
 Other
 ---------------------
 (No changes)

--- a/lucene/core/src/java/org/apache/lucene/index/SortingCodecReader.java
+++ b/lucene/core/src/java/org/apache/lucene/index/SortingCodecReader.java
@@ -314,6 +314,7 @@ public final class SortingCodecReader extends FilterCodecReader {
 
     SortingFloatVectorValues(FloatVectorValues delegate, Sorter.DocMap sortMap) throws IOException {
       this.delegate = delegate;
+      assert delegate != null;
       // SortingValuesIterator consumes the iterator and records the docs and ord mapping
       iteratorSupplier = iteratorSupplier(delegate, sortMap);
     }
@@ -446,6 +447,9 @@ public final class SortingCodecReader extends FilterCodecReader {
   @Override
   public FieldsProducer getPostingsReader() {
     FieldsProducer postingsReader = in.getPostingsReader();
+    if (postingsReader == null) {
+      return null;
+    }
     return new FieldsProducer() {
       @Override
       public void close() throws IOException {
@@ -481,6 +485,9 @@ public final class SortingCodecReader extends FilterCodecReader {
   @Override
   public StoredFieldsReader getFieldsReader() {
     StoredFieldsReader delegate = in.getFieldsReader();
+    if (delegate == null) {
+      return null;
+    }
     return newStoredFieldsReader(delegate);
   }
 
@@ -526,6 +533,9 @@ public final class SortingCodecReader extends FilterCodecReader {
   @Override
   public PointsReader getPointsReader() {
     final PointsReader delegate = in.getPointsReader();
+    if (delegate == null) {
+      return null;
+    }
     return new PointsReader() {
       @Override
       public void checkIntegrity() throws IOException {
@@ -551,6 +561,9 @@ public final class SortingCodecReader extends FilterCodecReader {
   @Override
   public KnnVectorsReader getVectorReader() {
     KnnVectorsReader delegate = in.getVectorReader();
+    if (delegate == null) {
+      return null;
+    }
     return new KnnVectorsReader() {
       @Override
       public void checkIntegrity() throws IOException {
@@ -587,6 +600,9 @@ public final class SortingCodecReader extends FilterCodecReader {
   @Override
   public NormsProducer getNormsReader() {
     final NormsProducer delegate = in.getNormsReader();
+    if (delegate == null) {
+      return null;
+    }
     return new NormsProducer() {
       @Override
       public NumericDocValues getNorms(FieldInfo field) throws IOException {
@@ -609,6 +625,9 @@ public final class SortingCodecReader extends FilterCodecReader {
   @Override
   public DocValuesProducer getDocValuesReader() {
     final DocValuesProducer delegate = in.getDocValuesReader();
+    if (delegate == null) {
+      return null;
+    }
     return new DocValuesProducer() {
       @Override
       public NumericDocValues getNumeric(FieldInfo field) throws IOException {
@@ -710,6 +729,9 @@ public final class SortingCodecReader extends FilterCodecReader {
   }
 
   private TermVectorsReader newTermVectorsReader(TermVectorsReader delegate) {
+    if (delegate == null) {
+      return null;
+    }
     return new TermVectorsReader() {
       @Override
       public void prefetch(int doc) throws IOException {

--- a/lucene/core/src/test/org/apache/lucene/index/TestSortingCodecReader.java
+++ b/lucene/core/src/test/org/apache/lucene/index/TestSortingCodecReader.java
@@ -153,14 +153,16 @@ public class TestSortingCodecReader extends LuceneTestCase {
         docIds.add(i);
       }
       Collections.shuffle(docIds, random());
-      // If true, index a vector for every doc
-      boolean denseVectors = random().nextBoolean();
+      // If true, index a vector and points for every doc
+      boolean dense = random().nextBoolean();
       try (RandomIndexWriter iw = new RandomIndexWriter(random(), dir)) {
         for (int i = 0; i < numDocs; i++) {
           int docId = docIds.get(i);
           Document doc = new Document();
           doc.add(new StringField("string_id", Integer.toString(docId), Field.Store.YES));
-          doc.add(new LongPoint("point_id", docId));
+          if (dense || docId % 3 == 0) {
+            doc.add(new LongPoint("point_id", docId));
+          }
           String s = RandomStrings.randomRealisticUnicodeOfLength(random(), 25);
           doc.add(new TextField("text_field", s, Field.Store.YES));
           doc.add(new BinaryDocValuesField("text_field", new BytesRef(s)));
@@ -172,7 +174,7 @@ public class TestSortingCodecReader extends LuceneTestCase {
           doc.add(new BinaryDocValuesField("binary_dv", new BytesRef(Integer.toString(docId))));
           doc.add(
               new SortedSetDocValuesField("sorted_set_dv", new BytesRef(Integer.toString(docId))));
-          if (denseVectors || docId % 2 == 0) {
+          if (dense || docId % 2 == 0) {
             doc.add(new KnnFloatVectorField("vector", new float[] {(float) docId}));
           }
           doc.add(new NumericDocValuesField("foo", random().nextInt(20)));
@@ -272,7 +274,7 @@ public class TestSortingCodecReader extends LuceneTestCase {
               assertTrue(sorted_numeric_dv.advanceExact(idNext));
               assertTrue(sorted_set_dv.advanceExact(idNext));
               assertTrue(binary_sorted_dv.advanceExact(idNext));
-              if (denseVectors || prevValue % 2 == 0) {
+              if (dense || prevValue % 2 == 0) {
                 assertEquals(idNext, valuesIterator.advance(idNext));
                 graph.seek(0, valuesIterator.index());
                 assertNotEquals(DocIdSetIterator.NO_MORE_DOCS, graph.nextNeighbor());
@@ -289,7 +291,7 @@ public class TestSortingCodecReader extends LuceneTestCase {
               assertEquals(1, sorted_numeric_dv.docValueCount());
               assertEquals(ids.longValue(), sorted_numeric_dv.nextValue());
 
-              if (denseVectors || prevValue % 2 == 0) {
+              if (dense || prevValue % 2 == 0) {
                 float[] vectorValue = vectorValues.vectorValue(valuesIterator.index());
                 assertEquals(1, vectorValue.length);
                 assertEquals((float) ids.longValue(), vectorValue[0], 0.001f);
@@ -306,9 +308,13 @@ public class TestSortingCodecReader extends LuceneTestCase {
                   leaf.storedFields().document(idNext).get("string_id"));
               IndexSearcher searcher = new IndexSearcher(r);
               TopDocs result =
-                  searcher.search(LongPoint.newExactQuery("point_id", ids.longValue()), 1);
-              assertEquals(1, result.totalHits.value());
-              assertEquals(idNext, result.scoreDocs[0].doc);
+                  searcher.search(LongPoint.newExactQuery("point_id", ids.longValue()), 10);
+              if (dense || ids.longValue() % 3 == 0) {
+                assertEquals(1, result.totalHits.value());
+                assertEquals(idNext, result.scoreDocs[0].doc);
+              } else {
+                assertEquals(0, result.totalHits.value());
+              }
 
               result =
                   searcher.search(new TermQuery(new Term("string_id", "" + ids.longValue())), 1);

--- a/lucene/core/src/test/org/apache/lucene/index/TestSortingCodecReader.java
+++ b/lucene/core/src/test/org/apache/lucene/index/TestSortingCodecReader.java
@@ -25,6 +25,8 @@ import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
 import java.util.Locale;
+
+import org.apache.lucene.codecs.KnnVectorsReader;
 import org.apache.lucene.codecs.TermVectorsReader;
 import org.apache.lucene.codecs.hnsw.HnswGraphProvider;
 import org.apache.lucene.document.BinaryDocValuesField;
@@ -247,8 +249,13 @@ public class TestSortingCodecReader extends LuceneTestCase {
             SortedSetDocValues sorted_set_dv = leaf.getSortedSetDocValues("sorted_set_dv");
             SortedDocValues binary_sorted_dv = leaf.getSortedDocValues("binary_sorted_dv");
             FloatVectorValues vectorValues = leaf.getFloatVectorValues("vector");
-            HnswGraph graph =
-                ((HnswGraphProvider) ((CodecReader) leaf).getVectorReader()).getGraph("vector");
+            KnnVectorsReader vectorsReader = ((CodecReader) leaf).getVectorReader();
+            HnswGraph graph;
+            if (vectorsReader instanceof HnswGraphProvider hnswGraphProvider) {
+              graph = hnswGraphProvider.getGraph("vector");
+            } else {
+              graph = null;
+            }
             NumericDocValues ids = leaf.getNumericDocValues("id");
             long prevValue = -1;
             boolean usingAltIds = false;
@@ -276,8 +283,10 @@ public class TestSortingCodecReader extends LuceneTestCase {
               assertTrue(binary_sorted_dv.advanceExact(idNext));
               if (dense || prevValue % 2 == 0) {
                 assertEquals(idNext, valuesIterator.advance(idNext));
-                graph.seek(0, valuesIterator.index());
-                assertNotEquals(DocIdSetIterator.NO_MORE_DOCS, graph.nextNeighbor());
+                if (graph != null) {
+                  graph.seek(0, valuesIterator.index());
+                  assertNotEquals(DocIdSetIterator.NO_MORE_DOCS, graph.nextNeighbor());
+                }
               }
 
               assertEquals(new BytesRef(ids.longValue() + ""), binary_dv.binaryValue());

--- a/lucene/core/src/test/org/apache/lucene/index/TestSortingCodecReader.java
+++ b/lucene/core/src/test/org/apache/lucene/index/TestSortingCodecReader.java
@@ -25,7 +25,6 @@ import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
 import java.util.Locale;
-
 import org.apache.lucene.codecs.KnnVectorsReader;
 import org.apache.lucene.codecs.TermVectorsReader;
 import org.apache.lucene.codecs.hnsw.HnswGraphProvider;

